### PR TITLE
[BEAM-14213] Add API and construction time validation for Batched DoFns

### PR DESCRIPTION
**Pull request by TheNeuralBit**
Split from #17175. This includes just the API, not the worker changes necessary to actually execute a Batched DoFn.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/python/apache_beam/typehints/batch.py
Comment (TheNeuralBit) [inline sdks/python/apache_beam/typehints/batch.py:102]: TODO: Remove BatchTypeConstraint concept. This is from a previous iteration of the Batched DoFn concept.

File: sdks/python/apache_beam/typehints/batch_test.py
Comment (TheNeuralBit) [inline sdks/python/apache_beam/typehints/batch_test.py:67]: TODO: verify DataFrame without index

Comment (TheNeuralBit) [reply sdks/python/apache_beam/typehints/batch_test.py:67]: Alternatively, we can just hold off on DataFrame batching support.

Comment (TheNeuralBit) [reply sdks/python/apache_beam/typehints/batch_test.py:55]: I backed out the pandas DataFrame batchConverter (f909d92). Branch https://github.com/TheNeuralBit/beam/tree/batched-dofn-pandas has this change added back

**Comment by TheNeuralBit** [comment]
R: @robertwb

File: sdks/python/apache_beam/typehints/batch.py
Comment (robertwb) [inline sdks/python/apache_beam/typehints/batch.py:66]: I know it's likely messy, but this parameter would benefit from type hints (and maybe a docstring).

File: sdks/python/apache_beam/typehints/batch.py
Comment (robertwb) [inline sdks/python/apache_beam/typehints/batch.py:73]: Should the API be to throw (possibly with an explanatory message) rather than return None?

File: sdks/python/apache_beam/typehints/batch.py
Comment (robertwb) [inline sdks/python/apache_beam/typehints/batch.py:94]: ListBatchConverter would be a useful one to have (both in practice, and as a canonical example).

File: sdks/python/apache_beam/utils/windowed_value.pxd
Comment (robertwb) [inline sdks/python/apache_beam/utils/windowed_value.pxd:46]: Let's make this an interface. (Maybe split it out into a different PR as well.)

Comment (robertwb) [reply sdks/python/apache_beam/utils/windowed_value.pxd:46]: In particular, for non-1:1 batch processing, we may want to (optionally?) expose the windowing information as part of the batch itself, e.g. as additional columns (for a dataframe-like batch), or as a list of WindowedValues (for a List batch), or ???

An alternative is to have batches of homogeneous windowing information (which would allow dropping the 1:1 requirement, at the expense of possibly smaller batches).

**Review by robertwb — COMMENTED** [review]
Great to see some concrete code. Only minor comments.

File: sdks/python/apache_beam/transforms/core.py
Comment (TheNeuralBit) [reply sdks/python/apache_beam/transforms/core.py]: Yes, thanks. I noticed this one while working on the worker changes.

File: sdks/python/apache_beam/transforms/batch_dofn_test.py
Comment (TheNeuralBit) [reply sdks/python/apache_beam/transforms/batch_dofn_test.py:41]: I wrote this up [here](https://docs.google.com/document/d/1Xt7T0e_TEJrCCx9DnUBY60zrysRHDw6mPBAcc-e9UHo/edit#). This was a useful exercise, but in the end I'm inclined to keep it as-is (although I'm open to changing it if a strong case is made). I'll send this out to the dev list.

File: sdks/python/apache_beam/utils/windowed_value.pxd
Comment (TheNeuralBit) [reply sdks/python/apache_beam/utils/windowed_value.pxd]: Ack I removed this from a PR. What does it mean to make it an interface? Just drop the `@cython.final`? (sorry my knowledge of cython is very limited :sweat_smile:)

File: sdks/python/apache_beam/typehints/batch.py
Comment (TheNeuralBit) [reply sdks/python/apache_beam/typehints/batch.py]: Added typehint

File: sdks/python/apache_beam/typehints/batch.py
Comment (TheNeuralBit) [reply sdks/python/apache_beam/typehints/batch.py:131]: Good point, done.

File: sdks/python/apache_beam/transforms/core.py
Comment (TheNeuralBit) [reply sdks/python/apache_beam/transforms/core.py]: Done - BEAM-14293

File: sdks/python/apache_beam/transforms/batch_dofn_test.py
Comment (TheNeuralBit) [reply sdks/python/apache_beam/transforms/batch_dofn_test.py]: These are meant to be expected values for the below parameterized test. Using a NamedTuple here was a bad idea, I switched the logic to just use dictionaries as they intended.

I might be missing your point though

File: sdks/python/apache_beam/transforms/batch_dofn_test.py
Comment (TheNeuralBit) [reply sdks/python/apache_beam/transforms/batch_dofn_test.py]: I broke this out as a free function

**Review by TheNeuralBit — COMMENTED** [review]
Thanks @robertwb, this still needs a little work but it's ready for another review.

File: sdks/python/apache_beam/typehints/batch.py
Comment (robertwb) [reply sdks/python/apache_beam/typehints/batch.py:75]: Yes. I would rather be too verbose than not verbose enough in the case of failure (e.g. a common error could be simply that a converter wasn't registered, so letting it say "doesn't fit" would still be valuable information).

**Review by robertwb — APPROVED** [review]
I'll follow up on the yields batches on the list, but I don't think this should block this PR getting in.

**Comment by robertwb** [comment]
For some reason I can't reply to comments, but regarding a BatchedWindowedValue "interface" I mean a just cdef class with unimplemented methods, not anything technical.

File: sdks/python/apache_beam/typehints/batch.py
Comment (TheNeuralBit) [reply sdks/python/apache_beam/typehints/batch.py:75]: Ok, I'd like to go ahead and get this in so we can focus on the worker changes now, so I filed BEAM-14339 for this.

File: sdks/python/apache_beam/transforms/core.py
Comment (apilloud) [inline sdks/python/apache_beam/transforms/core.py:716]: s/reqired/required/?

Comment (TheNeuralBit) [reply sdks/python/apache_beam/transforms/core.py:716]: thank you! I put in a superfluous apostrophe too :man_facepalming: 

https://github.com/apache/beam/pull/17452

**Pull request merged**
